text/0000-portals.md

```diff
@@ -0,0 +1,110 @@
+
+Having a custom element be inserted out of its natural dom tree hierarchy would enable one to have greater control over the stacking context, allowing a predictalbe popup and modal implementation (including faux-modals that are scoped to tabs and are not per-se global).
+
+Side-note: Modal scoping means that a modal would be tied to a tab (the backdrop only appears inside the tab, and the modal is centered against the tab content) and only blocks the content of the tab, and nothing outside of it (focus trapping still works as with a true modal, but there is a key that allows one to escape the modal). When using Lightning Console in Salesforce, all modals by default are scoped, so this is a very important use-case to capture.
+
+## Basic example
+
+The proposal suggests one new API -- a way to define the destination of an element (position in the DOM where the component would be rendered in). The api in question is a new attribute -- `lwc:portal` that accepts a string that should match the global document id of the element that would contain the rendered element.
+
+One non-trivial example of such usage:
+
+```xhtml
+<body>
+  <div id="global-modals"></div>
+
+  <div id="popups"></div>
+  ...
+  <div id="tabs">
+    <div id="tab-1">
+      <div id="tab1-modals"></div>
+
+      <div>
+         <x-dialog lwc:portal="global-modals">Global Modal</x-foo>
+         <x-dialog lwc:portal="tab1-modals">Scoped Modal</x-foo>
+      </div>  
+    </div>  
+    <div id="tab-2">
+      <div id="tab2-modals"></div>
+
+      <div>
+         <x-dialog lwc:portal="global-modals">Global Modal</x-foo>
+         <x-dialog lwc:portal="tab2-modals">Scoped Modal</x-foo>
+      </div>  
+    </div>  
+  </div>
+</body>
+```
+since we wouldn't want for a user to specify the portal each and every time one would use a dialog, more realistically -- `<x-dialog>` would have a default behaviour based on the `@Context`, which can provide whether scoping is enabled or not, the name of the scoped portal and the name of the global portal. This dynamic choice would indicate that support for dynamically specifying the portal may be needed.
+
+## Detailed design
+
+(not an actual detailed design, just adding some possible requirements)
+
+- If a custom element has `lwc:portal` attribute on it, and its value is an existing global id, instead of being inserted at its natural position in the template, the element would be inserted in the element with the defined global id.
+- When an element's destination portal doesn't exist, but later is inserted in the DOM, any elements that have already rendered stay in their original location, as if the portal doesn't exist.
+- When the host element is disconnected, the portalled element should also be disconnected. 
+- The portalled component should still be `@Context` aware 
```

Comment:
The current implementation for serving context to wire adapters uses custom DOM events. Therefore, without work there, context wire adapters will not receive context through their host source but through their portal container's context. It's possible to change this to support serving context through the source host instead of from the portal's context, but this needs to be debated whether or not that is correct.

I tend to think that context should be provided through the DOM composition, not the host source. It would still be possible to do this differently through manual context passing from host->portalled component, but doing it by default may be problematic.